Make TaskLevel immutable

jml opened the pull request:
Make `TaskLevel` immutable, moving the mutating part up to `Action`, which now knows what the last task it logged was.

Also adds some tests that use `hypothesis`. This slows the test suite a bit (~0.4s on my laptop), because although it looks like I'm adding two tests I'm actually adding two hundred. I'm happy to ditch hypothesis and rewrite them the normal way.

Although not strictly necessary for the parsing work I have planned in #188, this at least helps my thinking about the problem.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/clusterhq/eliot/189)
<!-- Reviewable:end -->

itamarst commented on eliot/_action.py:
```diff
@@ -106,19 +101,41 @@ def toString(self):
-    def nextChild(self):
+    def next(self):
+        """
+        Return the next L{TaskLevel}.
```
Next sibling? "next" by itself is somewhat ambiguous.

itamarst commented:
Huh. Why is Python 3 test run so much faster?

itamarst commented:
I'm happy to use `hypothesis`, it's a useful tool and I want to learn more about using it... and eliot is a good place to try it.

itamarst commented:
My guess is the slowdown is because the tests run with coverage tracking, and it's probably better implemented in Python 3. So disabling coverage testing would also make hypothesis have less of an impact, but lose information...

In practice is the test suite is fast enough that I don't care, but this might come up as an issue elsewhere.

jml commented:
I've renamed `next` to `next_sibling`. 

Alas, I also discovered that `TaskLevel` as `PClass` isn't ordered. I've added tests & basic comparison methods, as well as filed tobgu/pyrsistent#45. Please take a look.

itamarst commented:
`PClass` being non-ordered is deliberate, to be compatible with Python 3.

jml commented:
That seems like a non sequitur to me, but perhaps best discussed on the linked issue.

itamarst commented on eliot/_action.py:
```diff
@@ -94,7 +104,7 @@ def fromString(cls, string):
         @return: L{TaskLevel} parsed from the string.
         """
-        return cls(level=[int(i) for i in string.split("/") if i])
+        return cls(level=pvector([int(i) for i in string.split("/") if i]))
```
Not actionable, but you don't need need the `pvector`; `pvector_field` will do taht for you).

jml replied:
Done.

itamarst commented on eliot/_action.py:
```diff
@@ -106,19 +116,42 @@ def toString(self):
+        return TaskLevel(level=self.level.set(-1, self.level[-1] + 1))
+
+
+    def child(self, level=1):
```
Argument is never used, so maybe just drop it.

itamarst replied:
Or rather, it's not ever used in actual code, only by tests, so better to have tests match the way actual code uses `TaskLevel`?

jml replied:
`level` param dropped.

itamarst commented on eliot/tests/test_action.py:
```diff
@@ -791,26 +794,84 @@ def test_continueTaskRequiredTaskId(self):
-                         [TaskLevel(level=[1]), TaskLevel(level=[2]),
-                          TaskLevel(level=[3, 1]), TaskLevel(level=[3, 2]),
-                          TaskLevel(level=[4])])
+        self.assertTrue(all([
```
Python! It's the best.

itamarst commented on eliot/_action.py:
```diff
@@ -171,6 +204,7 @@ def __init__(self, logger, task_uuid, task_level, action_type,
                  DeprecationWarning, stacklevel=2)
             task_level = TaskLevel.fromString(task_level)
         self._task_level = task_level
+        self._last_child = None
```
:question: An alternate thing you could do is store `_next_child`, and then the type of the attribute stays consistent.

jml replied:
Acknowledged.

itamarst commented:
OK, there's a bunch of minor comments; address to your best judgement and then merge.
